### ISSUE TITLE
let's think about fonts

### ISSUE THREAD

**Issue by lastquestion** [opentable/design-tokens#26]
## values for fonts
So, we need to support a variety of values for fonts. We need to transport

* font weights
* font sizes
* line heights
* font family

All of these are fine to transport via values.

So in particular, I think we can be pretty low-level here in design-token land. I would imagine something like this: (values made up, names made up, etc.):

```
regular-weight: 200;
medium-line-height: 1rem;
medium-size: 0.875rem;
regular-font-family: 'source-sans, arial';
```

(One little thing here. I like the idea of not being too verbose, so since we're already in fonts, why do we need to throw font in the names of stuff everywhere? e.g. `medium-size` versus `medium-font-size`.)

## Not encoding @font-face or where the fonts are located
I would like not to deal with `@font-face` here. It's a platform specific problem of web, and doesn't apply to native platforms (which we care about, in the future). Additionally, I'd like to ***not*** encode the way that the font is found in the design token.  

So I will open another issue in react land to deal with that - how to encode `@font-face`, and where the font is actually loaded. There is some interaction with `:global` and CSS-modules (bugs, basically, in webpack's `css-loader`). I think we also need to consider that `@font-face` needs to be a global rule, and cannot be a `local` in a CSS module. 

## how to encode text styles?
From a design perspective, we absolutely have "groupings" of "font design tokens" that define a "text style" (I made these words up just now). But the way to encode this becomes interesting. 

We want to define the "text style name", and the "attribute values for the font style", in a non-platform specific way. Therefore, we should probably encode them like this:

```
text-style-1:
  weight: "{!regular-weight}
  size: "{!medium-size}"
  line-height: "{!medium-line-height}"
  font-family: "{!brand-family}"

text-style-2:
  ....
```

A couple of points here:

* no, those names do not match CSS names, on purpose. platform agnostic, yo.
* Alias probably works ok here, since when we change the alias value, it will change everywhere
* I used `text-style-1` to avoid discussions about `h1`-`h4`. 👼 

The idea is we extend `theo` to generate correct representations of this in each target format. For example, in CSS-module, I'd propose we generate:

```
@value text-style-1: "200 0.875rem/1rem source-sans, arial"
```

(I may have got the syntax slightly off). In other words, that is, provide a singular **value** that later we can import as

```
font: @text-style-1
```

In JS, I propose we'd just provide a hash that maps 1-1. I guess we need to decide if the JS actually is a CSS object and if so, we may have to munge the prop names. As an aside, it's quite interesting. Is JS supposed to be a representation of the `design-tokens` or is it supposed to be matching some CSS-in-JS standard? ;)

This way, other platforms would be able to generate the right artifact for their platform from the base values themselves.

## text-styles from a design perspective

I think most of the time, designers *should* be directly using one of the defined `text-style`. But sometimes, it may be right - and designers should decide this - to use values directly from the font design token. In which case, they are all available here too.

One of the side effects of actually writing out the text-styles as something aliasable is that we need to be really, really, careful about the naming and intent of these text-styles. 

They should represent a "design object" that designers can pull from their toolkit and use with the exact same **semantic** and **visual** meaning. When we go to update the values in a `text-style`, everywhere that text-style is used will be updated, and it is important to realize that is ***by design***. This will be the biggest area of "unintended change" that is possible. 

For example, if a designer notices he plans on using something that ***visually matches*** `caption-text`, but not for something that sematically means a ***caption***, then he should manually define out the fonts, or create a new text-style. Because later, if we decide to change what a `caption` looks like across the site, this text will also change, unintended.

This is something that design & design review will need to manage from a overall design.

And, of course adding new text-styles would absolutely be an area where design needs to sign off on the PR.

cc @lindseyhight , @mannionaco , @francesyun 
cc @Kimtaro 
cc @ericz82 , @chriscartlidge

**Comment by lastquestion** [comment]
Although, maybe we should actually store the font (check it in) here, so that platforms can read it out and then upload/incorporate into bundle/whatever. I think that's probably a good idea. But are we allowed to check in fonts if it's OSS? I don't know the legalese around our fonts. Probably if we buy them? I know source sans is typekit?

**Comment by Kimtaro** [comment]
I like the idea of text styles 👍 

We also need to account for switching `font-family` and `line-height` by the `lang` attribute, as we do in opentable-theme: https://github.com/opentable/opentable-theme/blob/master/assets/css/_settings.scss#L79-L85

**Comment by nickbalestra** [comment]
@lastquestion thanks a lot for all the points.
Here my add (that also addresses @Kimtaro points):

I wouldn't add specific format logic in the token itselves (as per your cssmodule suggestion) and instead promote logic/assembling at consumer-point.

I'll suggest to groups thing within those big categories (as those should be the things that a typography token need to address):

### base
- font family base (ie: "''Helvetica Neue', 'Arial', sans-serif")
- font color base
- font size base
- font weight bold base
- font letter spacing base
- font size root (, ie: "100%")
### text styles
(I like to stay away of h1-h6, what about aligning to what we are doing with spacing..)
- text size xsmall
- text size small
- text size medium
- text size large
- text size xlarge
- text weight xsmall
- text weight small
- text weight medium
- text weight large
- text letter spacing xsmall
- text letter spacing small
- text letter spacing medium
- text letter spacing large
- text letter spacing xlarge
### Line heights
- line height xsmall
- line height small
- line height medium
- line height large
- line height xlarge
### headings
- heading margin top
- heading margin bottom
### paragraphs
- p margin top
- p margin bottom
### lists
- list margin top
- list margin bottom
- list item margin top
- list item margin bottom
### links
- link color
- link hover color
- link text decoration
- link hover coor
- link active color
- link visited color
### Lang
- font family --lang

Another good example is [this](https://github.com/Skyscanner/backpack/blob/master/packages/bpk-tokens/src/base/typography.json).

@lindseyhight @mannionaco what you recon?

**Comment by lastquestion** [comment]
@nickbalestra 

### about the idea of base - well I mean philosophically
Looking at the typography invision link, one thing I suggested was, from a design perspective, "there is no base style". 

EG: instead of thinking of "base" + "overrides to get to x", rather just think "x is made of < many small things >".

I'd like to try and carry that forward over to the tokens themselves by sort of removing the idea of "base". I in reality mean from the words and names of the tokens and matching the names that design is using. 

### specific thoughts on the values you proposed
Overall, yeah, :+1:, that looks complete. Some thoughts:
1. Although there are font-sizes of `xsmall, small, medium, large, xlarge`, there isn't so many font weights or letter spacings. We won't need so many.
2. `margin` - I think the margin is part of design system. That is the spacing between paragraphs is defined by the spacing system. Therefore from a design perspective there is no `p`, no `ul`. They have no margins, either. We don't need these in the typography token set because the component with the actual text will set the margin to either 0 or whatever the design specifies from the spacing token set (small, medium, etc.) 
3. About colors: I think we do not need an alias here from link -> color system. Design should mostly specify things like "`link` here", and the link component should take the text-style and add the color which would be `primary-red` or whatnot, and add the hover color, etc. OK, I know this is going to sound crazy, so I welcome your reply on this one. 😁 . My claim: there's no need to alias because the alias happens at the design level, and implemented at the component level. So we don't need a SCSS style $token->$token->final value. In fact, I feel that actually makes things more complicated. If later we need to change the color or margin or whatever of links, we can do it by changing the component itself, not by realiasing the token. In some more abstract sense, this is proposing moving as much of the logic into JS as possible, instead of putting logic in the alias/CSS world. 

### about font weights
I think we either use numbers or `bold`, `semibold`, `normal`. I don't know what design has picked - there might be some thing we need to think about with fallback fonts? I think font weights will try to fallback to the closet available font weight, but we should investigate more. 

I don't see that we need `text-weight xsmall, small, large` in general, though. We only have three in the spec - `bold`, `regular`, and `medium` (using the words from the invision board here, not CSS terms).

### about multilanguage
@Kimtaro I think originally people were thinking we can use one font-family and define the language fallbacks in there all together. Does that work? If not, we need to think about how to do this. Also, no one has considered line-height adjustments for languages. If we do that, I think first design needs to think through the implications.  cc @mannionaco 

Implementation wise, it's also a bit problematic. Consider that we won't be using SCSS to apply rules, and the current suggestion I have doesn't use `@import` either. How do you think we should pick what lang we're in? 

I would prefer to be "standards-compliant" and use the `lang` attribute on the HTML and then use the psuedo-selector `:lang`, but we need to inject that **everywhere we define a text-style** which would be in each specific seletor in many different CSS modules. 😢  (For example). And what about other export formats? like `js`, etc. We will need to encode the rules in YAML so that they can be generated into each format correctly.

For CSS, I guess we should consider instead of transporting text-style via a `font` attribute, using `compose` to import a set of CSS rules instead. I am thinking it will work, but it's so much simpler to just directly use a `font` attribute. 

I guess, that usually we do such lang font stuff via global selectors, and I want to think about how to try to not do that and also at least consider other platforms :)

**Comment by Kimtaro** [comment]
# About multi-language

(Sorry for hijacking the thread with a ton of language stuff 😄)

## Single font-family for all languages

There's two problems with this approach.

Japanese text that contains English characters and numbers is very common, so if we define Western fonts before the Japanese fonts we would end up rendering the English text in those fonts instead of the Japanese fonts. That itself is not an issue, and is in fact a common practice in Japanese typography, but it's not what we're doing now, so we'd have to get design to make sure that the Western/Japanese font-pairing would look and function well together.

The other issue is when we want to add Chinese language support. Whichever of Chinese and Japanese would come first in the font stack would end up rendering both languages, regardless of the `lang` attribute. https://codepen.io/kimtaro/pen/vJxYbY?editors=1100

## Using `:lang` pseudoselector

This is what we're doing now using `@mixin`: https://github.com/opentable/opentable-theme/blob/master/assets/css/_settings.scss#L79-L85

This is the nice upside of using SASS.

## Using `compose`

I'm not familiar enough with CSS modules yet to comment on this, but if it works I'm happy :)

## Language specific `line-height` adjustments

I explored this for Japanese in https://github.com/opentable/opentable-theme/pull/1441, but decided to defer until we got the new system in place. tl;dr: Japanese needs larger line-height than Western languages to look good.

**Comment by lastquestion** [comment]
cc @morewry

**Comment by lastquestion** [comment]
I'm going to try to use `compose` correctly with `:lang` in the react prototype github I did github.com/lastquestion/react-ssr-css-modules-values and see how it goes. Unfortunately, if we need to have different font stacks and `line-height`, I think we need to compose multiple rules in as opposed to a singular `font` attribute. 😢 

On the bright side, from what I understand of the CSS module `compose` functionality, it doesn't actually duplicate the rules a zillion times. 👍

**Comment by lastquestion** [comment]
So the following PR that @lindseyhight worked on #29 gave us the font sizes, weights, and so on.

This PR I just opened is a stab at grouping the text-styles together. I am pretty open to anything including completely different approaches to doing the grouping: https://github.com/opentable/design-tokens/pull/32

There's one or two interesting things about that proposal, I left some comments in the PR description.

Basically, it's transporting the token values for each text-style in it's own `.yml` file, and then providing an extra CSS module so that the text-style can be `compose:` easily. I actually tried to write a transform in `theo` to generate that automatically, and got somewhere, but pulled back because I'm not sure if this is the right direction to do.

I can also see a strong argument that that file doesn't even belong here. It could belong elsewhere, like in ot-react-ui. :man_shrugging: I can see that. It's kind of nice to put it here, because it's like, "part of the output format for CSS Modules for text-styles", though.
